src/dune_engine/build_system.ml

```diff
@@ -1584,6 +1584,25 @@ end = struct
                reason)
         ]
 
+  (* All actions that raised errors in [execute_rule_impl] depend on this cell.
+     By invalidating this cell, one can therefore force them to be rerun. *)
+  let error_cell = Memo.lazy_cell ~name:"error-cell" Memo.Build.return
+
+  let of_reproducible_fiber_with_error_cell_dependency fiber ~on_error =
+    let handled_error = ref false in
+    let on_error exn =
+      handled_error := true;
+      on_error exn
+    in
+    let fiber = Fiber.with_error_handler fiber ~on_error in
+    let* result = Memo.Build.of_non_reproducible_fiber fiber in
+    let+ () =
+      match !handled_error with
+      | true -> Memo.Cell.read error_cell
```

Comment:
This code looks unreachable: if we get an exception, `on_error` will raise and `with_error_handler` will never terminate.

Reply:
Oh, whoops, I don't know what I was thinking when I read this.

Reply:
Oh, no :( That's the embarrassment I get for skipping writing a test.

Reply:
Reverting in #4731